libbeat/monitoring/adapter/go-metrics.go

```diff
@@ -0,0 +1,165 @@
+package adapter
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/monitoring"
+	metrics "github.com/rcrowley/go-metrics"
+)
+
+// implement adapter for adding go-metrics based counters
+// to monitoring
+
+// GoMetricsRegistry wraps a monitoring.Registry for filtering and registering
+// go-metrics based metrics with the monitoring package.
+type GoMetricsRegistry struct {
+	reg     *monitoring.Registry
+	filters *metricFilters
+}
+
+func GetGoMetrics(parent *monitoring.Registry, name string, filters ...MetricFilter) *GoMetricsRegistry {
+	v := parent.Get(name)
+	if v == nil {
+		return NewGoMetrics(parent, name, filters...)
+	}
+
+	reg := v.(*monitoring.Registry)
+	return &GoMetricsRegistry{
+		reg:     reg,
+		filters: makeFilters(filters),
+	}
+}
+
+func NewGoMetrics(parent *monitoring.Registry, name string, filters ...MetricFilter) *GoMetricsRegistry {
+	return &GoMetricsRegistry{
+		reg:     parent.NewRegistry(name, monitoring.IgnorePublishExpvar),
+		filters: makeFilters(filters),
+	}
+}
+
+func (r *GoMetricsRegistry) Each(f func(string, interface{})) {
+	// empty, all metrics are mine
+}
+
+func (r *GoMetricsRegistry) find(name string) interface{} {
+	st := r.findState(name)
+	if st.action == actIgnore {
+		return nil
+	}
+
+	return r.reg.Get(st.name)
+}
+
+func (r *GoMetricsRegistry) Get(name string) interface{} {
+	m := r.find(name)
+	if m == nil {
+		return nil
+	}
+
+	if w, ok := m.(goMetricsWrapper); ok {
+		return w.wrapped()
+	}
+
+	return m
+}
+
+func (r *GoMetricsRegistry) GetOrRegister(name string, metric interface{}) interface{} {
+	v := r.Get(name)
+	if v != nil {
+		return v
+	}
+
+	return r.doRegister(name, metric)
+}
+
+func (r *GoMetricsRegistry) Register(name string, metric interface{}) error {
+	if r.find(name) != nil {
+		return fmt.Errorf("metric '%v' already registered", name)
+	}
+
+	r.doRegister(name, metric)
+	return nil
+}
+
+func (r *GoMetricsRegistry) doRegister(name string, metric interface{}) interface{} {
+	if v := reflect.ValueOf(metric); v.Kind() == reflect.Func {
+		metric = v.Call(nil)[0].Interface()
+	}
+
+	st := r.addState(name, metric)
+	if st.action == actAccept {
+		w, ok := goMetricsWrap(st.metric)
+		if ok {
+			r.reg.Add(st.name, w)
+		}
+	}
+
+	return st.metric
+}
+
+func (r *GoMetricsRegistry) RunHealthchecks() {}
```

Comment:
yes, purpose of `adapter` package is to integrate other metrics collection into `libbeat/monitoring`.